File: sdk/go/protomcp/group_test.go
package protomcp

import (
	"context"
	"strings"
	"testing"
)

func dummyToolContext() ToolContext {
	return ToolContext{Ctx: context.Background()}
}

func TestGroupRegistration(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("math",
		GroupDescription("Math operations"),
		Action("add",
			ActionDescription("Add two numbers"),
			ActionArgs(IntArg("a"), IntArg("b")),
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				return Result("sum")
			}),
		),
		Action("multiply",
			ActionDescription("Multiply two numbers"),
			ActionArgs(IntArg("x"), IntArg("y")),
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				return Result("product")
			}),
		),
	)

	groups := GetRegisteredGroups()
	if len(groups) != 1 {
		t.Fatalf("expected 1 group, got %d", len(groups))
	}
	if groups[0].Name != "math" {
		t.Errorf("expected name 'math', got '%s'", groups[0].Name)
	}
	if groups[0].Description != "Math operations" {
		t.Errorf("expected description 'Math operations', got '%s'", groups[0].Description)
	}
	if len(groups[0].Actions) != 2 {
		t.Errorf("expected 2 actions, got %d", len(groups[0].Actions))
	}
}

func TestUnionStrategySchema(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("db",
		GroupDescription("DB ops"),
		Action("query",
			ActionDescription("Run query"),
			ActionArgs(StrArg("sql")),
		),
		Action("insert",
			ActionDescription("Insert record"),
			ActionArgs(StrArg("table"), ObjectArg("data")),
		),
	)

	defs := GroupsToToolDefs()
	if len(defs) != 1 {
		t.Fatalf("expected 1 tool def, got %d", len(defs))
	}
	td := defs[0]
	if td.Name != "db" {
		t.Errorf("expected name 'db', got '%s'", td.Name)
	}

	schema := td.InputSchema
	props, _ := schema["properties"].(map[string]interface{})
	actionProp, _ := props["action"].(map[string]interface{})
	enumVals, _ := actionProp["enum"].([]interface{})
	if len(enumVals) != 2 {
		t.Fatalf("expected 2 enum values, got %d", len(enumVals))
	}

	oneOf, _ := schema["oneOf"].([]interface{})
	if len(oneOf) != 2 {
		t.Fatalf("expected 2 oneOf entries, got %d", len(oneOf))
	}

	// Verify first entry has action const and sql
	entry0 := oneOf[0].(map[string]interface{})
	entryProps0 := entry0["properties"].(map[string]interface{})
	actionConst0 := entryProps0["action"].(map[string]interface{})
	if actionConst0["const"] != "query" {
		t.Errorf("expected first action const 'query', got '%v'", actionConst0["const"])
	}
	if _, ok := entryProps0["sql"]; !ok {
		t.Error("expected 'sql' in query properties")
	}
}

func TestSeparateStrategySchema(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("files",
		GroupDescription("File ops"),
		GroupStrategy("separate"),
		Action("read",
			ActionDescription("Read a file"),
			ActionArgs(StrArg("path")),
		),
		Action("write",
			ActionDescription("Write a file"),
			ActionArgs(StrArg("path"), StrArg("content")),
		),
	)

	defs := GroupsToToolDefs()
	if len(defs) != 2 {
		t.Fatalf("expected 2 tool defs, got %d", len(defs))
	}
	names := map[string]bool{}
	for _, d := range defs {
		names[d.Name] = true
	}
	if !names["files.read"] {
		t.Error("expected 'files.read' tool def")
	}
	if !names["files.write"] {
		t.Error("expected 'files.write' tool def")
	}
}

func TestDispatchCorrectAction(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("calc",
		Action("add",
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				a := int(args["a"].(float64))
				b := int(args["b"].(float64))
				return Result(strings.Repeat("x", a+b))
			}),
			ActionArgs(IntArg("a"), IntArg("b")),
		),
	)

	groups := GetRegisteredGroups()
	ctx := dummyToolContext()
	result := DispatchGroupAction(groups[0], ctx, map[string]interface{}{
		"action": "add",
		"a":      float64(3),
		"b":      float64(4),
	})
	if result.IsError {
		t.Errorf("unexpected error: %s", result.ResultText)
	}
	if result.ResultText != "xxxxxxx" {
		t.Errorf("expected 7 x's, got '%s'", result.ResultText)
	}
}

func TestDispatchUnknownAction(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("calc2",
		Action("add",
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				return Result("ok")
			}),
		),
	)

	groups := GetRegisteredGroups()
	ctx := dummyToolContext()
	result := DispatchGroupAction(groups[0], ctx, map[string]interface{}{
		"action": "ad",
	})
	if !result.IsError {
		t.Error("expected error for unknown action")
	}
	if !strings.Contains(result.ResultText, "Unknown action") {
		t.Errorf("expected 'Unknown action' in result, got '%s'", result.ResultText)
	}
	if !strings.Contains(result.ResultText, "add") {
		t.Errorf("expected 'add' suggestion in result, got '%s'", result.ResultText)
	}
}

func TestDispatchMissingAction(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("calc3",
		Action("add",
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				return Result("ok")
			}),
		),
	)

	groups := GetRegisteredGroups()
	ctx := dummyToolContext()
	result := DispatchGroupAction(groups[0], ctx, map[string]interface{}{})
	if !result.IsError {
		t.Error("expected error for missing action")
	}
	if !strings.Contains(result.ResultText, "Missing") {
		t.Errorf("expected 'Missing' in result, got '%s'", result.ResultText)
	}
}

func TestGroupsInGetRegisteredTools(t *testing.T) {
	ClearRegistry()
	ClearGroupRegistry()
	defer func() {
		ClearRegistry()
		ClearGroupRegistry()
	}()

	ToolGroup("tools_test",
		GroupDescription("Test group"),
		Action("ping",
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				return Result("pong")
			}),
		),
	)

	tools := GetRegisteredTools()
	found := false
	for _, td := range tools {
		if td.Name == "tools_test" {
			found = true
			break
		}
	}
	if !found {
		t.Error("expected 'tools_test' in registered tools")
	}
}

func TestUnionHandlerDispatch(t *testing.T) {
	ClearGroupRegistry()
	defer ClearGroupRegistry()

	ToolGroup("handler_test",
		Action("greet",
			ActionArgs(StrArg("name")),
			ActionHandler(func(ctx ToolContext, args map[string]interface{}) ToolResult {
				return Result("Hello, " + args["name"].(string) + "!")
			}),
		),
	)

	defs := GroupsToToolDefs()
	ctx := dummyToolContext()
	result := defs[0].HandlerFn(ctx, map[string]interface{}{
		"action": "greet",
		"name":   "World",
	})
	if result.ResultText != "Hello, World!" {
		t.Errorf("expected 'Hello, World!', got '%s'", result.ResultText)
	}
}

File: sdk/typescript/tests/group.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { z } from 'zod';
import { toolGroup, getRegisteredGroups, clearGroupRegistry, groupsToToolDefs } from '../src/group.js';
import { getRegisteredTools, clearRegistry } from '../src/tool.js';
import { ToolContext } from '../src/context.js';

function dummyCtx(): ToolContext {
  return new ToolContext('', () => {});
}

beforeEach(() => {
  clearGroupRegistry();
  clearRegistry();
});

describe('toolGroup', () => {
  it('registers a group', () => {
    toolGroup({
      name: 'math',
      description: 'Math operations',
      actions: {
        add: {
          description: 'Add two numbers',
          args: z.object({ a: z.number(), b: z.number() }),
          handler: (args) => args.a + args.b,
        },
        multiply: {
          description: 'Multiply two numbers',
          args: z.object({ x: z.number(), y: z.number() }),
          handler: (args) => args.x * args.y,
        },
      },
    });

    const groups = getRegisteredGroups();
    expect(groups).toHaveLength(1);
    expect(groups[0].name).toBe('math');
    expect(groups[0].description).toBe('Math operations');
    expect(Object.keys(groups[0].actions)).toHaveLength(2);
  });

  it('generates union strategy schema', () => {
    toolGroup({
      name: 'db',
      description: 'DB ops',
      actions: {
        query: {
          description: 'Run query',
          args: z.object({ sql: z.string() }),
          handler: (args) => args.sql,
        },
        insert: {
          description: 'Insert record',
          args: z.object({ table: z.string(), data: z.object({}) }),
          handler: () => 'ok',
        },
      },
    });

    const defs = groupsToToolDefs();
    expect(defs).toHaveLength(1);
    const td = defs[0];
    expect(td.name).toBe('db');
    const schema = JSON.parse(td.inputSchemaJson);
    expect(new Set(schema.properties.action.enum)).toEqual(new Set(['query', 'insert']));
    expect(schema.oneOf).toHaveLength(2);

    const entries: Record<string, any> = {};
    for (const e of schema.oneOf) {
      entries[e.properties.action.const] = e;
    }
    expect(entries.query.properties.sql).toBeDefined();
    expect(entries.insert.properties.table).toBeDefined();
    expect(entries.insert.properties.data).toBeDefined();
  });

  it('generates separate strategy schema', () => {
    toolGroup({
      name: 'files',
      description: 'File ops',
      strategy: 'separate',
      actions: {
        read: {
          description: 'Read a file',
          args: z.object({ path: z.string() }),
          handler: (args) => args.path,
        },
        write: {
          description: 'Write a file',
          args: z.object({ path: z.string(), content: z.string() }),
          handler: () => 'ok',
        },
      },
    });

    const defs = groupsToToolDefs();
    expect(defs).toHaveLength(2);
    const names = defs.map((d) => d.name);
    expect(names).toContain('files.read');
    expect(names).toContain('files.write');

    const readDef = defs.find((d) => d.name === 'files.read')!;
    const readSchema = JSON.parse(readDef.inputSchemaJson);
    expect(readSchema.properties.path).toBeDefined();
    expect(readSchema.properties.action).toBeUndefined();
  });

  it('dispatches correct action via union handler', () => {
    toolGroup({
      name: 'calc',
      description: 'Calculator',
      actions: {
        add: {
          description: 'Add',
          args: z.object({ a: z.number(), b: z.number() }),
          handler: (args) => args.a + args.b,
        },
      },
    });

    const defs = groupsToToolDefs();
    const result = defs[0].handler({ action: 'add', a: 3, b: 4 }, dummyCtx());
    expect(result).toBe(7);
  });

  it('returns error for unknown action', () => {
    toolGroup({
      name: 'calc2',
      description: 'Calculator',
      actions: {
        add: {
          description: 'Add',
          args: z.object({ a: z.number(), b: z.number() }),
          handler: (args) => args.a + args.b,
        },
      },
    });

    const defs = groupsToToolDefs();
    const result = defs[0].handler({ action: 'ad' }, dummyCtx());
    expect(result.isError).toBe(true);
    expect(result.result).toContain('Unknown action');
    expect(result.result).toContain('add');
  });

  it('returns error for missing action', () => {
    toolGroup({
      name: 'calc3',
      description: 'Calculator',
      actions: {
        add: {
          description: 'Add',
          args: z.object({ a: z.number() }),
          handler: (args) => args.a,
        },
      },
    });

    const defs = groupsToToolDefs();
    const result = defs[0].handler({}, dummyCtx());
    expect(result.isError).toBe(true);
    expect(result.result).toContain('Missing');
  });

  it('groups appear in getRegisteredTools', () => {
    toolGroup({
      name: 'tools_test',
      description: 'Test group',
      actions: {
        ping: {
          description: 'Ping',
          args: z.object({}),
          handler: () => 'pong',
        },
      },
    });

    const tools = getRegisteredTools();
    const names = tools.map((t) => t.name);
    expect(names).toContain('tools_test');
  });

  it('dispatches separate strategy handlers', () => {
    toolGroup({
      name: 'sep_test',
      description: 'Sep test',
      strategy: 'separate',
      actions: {
        echo: {
          description: 'Echo',
          args: z.object({ msg: z.string() }),
          handler: (args) => args.msg,
        },
      },
    });

    const defs = groupsToToolDefs();
    const result = defs[0].handler({ msg: 'hi' }, dummyCtx());
    expect(result).toBe('hi');
  });
});
